src/systemctl/systemctl-edit.c

```diff
@@ -122,25 +126,74 @@ static int create_edit_temp_file(const char *new_path, const char *original_path
+                        return r;
 
-        r = copy_file(original_path, t, 0, 0644, 0, 0, COPY_REFLINK);
-        if (r == -ENOENT) {
+                r = copy_file(original_path, t, 0, 0644, 0, 0, COPY_REFLINK);
+                if (r == -ENOENT) {
+                        r = touch(t);
+                        mac_selinux_create_file_clear();
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to create temporary file \"%s\": %m", t);
+                } else {
+                        mac_selinux_create_file_clear();
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to create temporary file for \"%s\": %m", new_path);
+                }
+        } else if (original_unit_paths) {
+                _cleanup_free_ char *new_contents = NULL;
+                _cleanup_fclose_ FILE *f = NULL;
+                char **path;
+                size_t size;
 
-                r = touch(t);
+                r = mac_selinux_create_file_prepare(new_path, S_IFREG);
+                if (r < 0)
+                        return r;
 
+                f = fopen(t, "we");
                 mac_selinux_create_file_clear();
+                if (!f)
+                        return log_error_errno(errno, "Failed to open \"%s\": %m", t);
 
+                r = fchmod(fileno(f), 0644);
                 if (r < 0)
-                        return log_error_errno(r, "Failed to create temporary file \"%s\": %m", t);
-
-        } else {
-                mac_selinux_create_file_clear();
+                        return log_error_errno(errno, "Failed to change mode of \"%s\": %m", t);
+
+                r = read_full_file(new_path, &new_contents, &size);
+                if (r < 0 && r != -ENOENT)
+                        return log_error_errno(r, "Failed to read \"%s\": %m", new_path);
+
+                fprintf(f,
+                        "### Editing %s\n"
+                        EDIT_MARKER_START
+                        "\n\n%s%s\n"
+                        EDIT_MARKER_END
+                        "\n",
+                        new_path,
+                        strempty(new_contents),
+                        new_contents && endswith(new_contents, "\n") ? "" : "\n");
+
+                /* Add a comment with the contents of the original unit files */
+                STRV_FOREACH(path, original_unit_paths) {
+                        _cleanup_free_ char *contents = NULL;
+
+                        /* Skip the file that's being edited */
+                        if (path_equal(*path, new_path))
+                                continue;
+
+                        r = read_full_file(*path, &contents, &size);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to read \"%s\": %m", *path);
+
+                        fprintf(f, "\n# %s\n", *path);
+                        fwrite(contents, sizeof(char), strlen(contents), f);
```

Comment:
hmm, wouldn't it make sense to prefix the lines of the drop-ins here too? you copy them verbatim here, but that's confusing no?

```c
strreplace(contents, "\n", "\n# ");
```
should already suffice for that, no?

Reply:
Hopefully the marker headers are clear enough to explain how it works, but yeah, commenting out the contents is an option and would be clearer. Keeping it uncommented gives you syntax highlighting and makes it easier to copy stuff.
Would you prefer the existing units be commented out then?

Reply:
Yeah i think so. i mean syntax highlighting then hopefully grays it out, which I think is actually desirable, given it actually is just informational and won't have effect. If informational stuff without effect is gray, that's inherently a *good* thing I would claim.

(I mean, ideally we'd have "desatured" version of the regular colorful syntax highlighting in our text editors for cases like this, but if the choice is between totally greyed out and totally colorful syntax hightlighting for this informational stuff I'd argue for the former)

Reply:
That's sensible - I've applied those adjustments